### PR TITLE
Profile edit

### LINKED ISSUE #145: Profile edit
Allow the auth'd user to edit their profile. There should be some kind of edit button when viewing the profile that allows them to change out all of the fields. We'll need a PUT endpoint, but the edit form could be worked on in isolation.

### PR DESCRIPTION
**Pull request by AMStrix**
Closes #145.

### What this does
- Profile edit UI
- basic backend `/users/<id>` PUT update functionality with tests

### What this doesn't do
- backend user auth check
- custom backend field validation 
- image upload
- borrows heavily from create flow TeamMember edit, but differs enough for me to hesitate DRYing them together, I suspect they will diverge further in the future

### Test
- visit any user other than authenticated user and ensure "Edit profile" doesn't appear
- visit authenticated user profile, click "Edit profile" button
- observe that the "Save changes" button only becomes enabled when something has been changed
- check that all editable fields update on save
- try adding image and deleting image

### Screencaps
screen shot 2018-10-22 at 10 55 34 pm
screen shot 2018-10-22 at 10 55 53 pm
screen shot 2018-10-22 at 10 56 53 pm

### CODE REVIEW

File: backend/grant/user/views.py [hunk 1]
```diff
@@ -74,3 +74,36 @@ def create_user(account_address, email_address, display_name, title):
...
+
+@blueprint.route("/<user_identity>", methods=["PUT"])
+@endpoint.api(
+    parameter('accountAddress', type=str, required=True),
+    parameter('emailAddress', type=str, required=True),
+    parameter('displayName', type=str, required=True),
+    parameter('title', type=str, required=True),
+    parameter('socialMedias', type=list, required=True),
+    parameter('avatar', type=dict, required=False)
+)
```
Comment (wbobeirne) [inline backend/grant/user/views.py]: I think no params should be required, and we should only update the fields that were passed.

File: backend/grant/user/views.py [hunk 2]
```diff
@@ -74,3 +74,36 @@ def create_user(account_address, email_address, display_name, title):
 
     result = user_schema.dump(user)
     return result
+
+
+@blueprint.route("/<user_identity>", methods=["PUT"])
+@endpoint.api(
+    parameter('accountAddress', type=str, required=True),
+    parameter('emailAddress', type=str, required=True),
```
Comment (wbobeirne) [inline backend/grant/user/views.py]: If they can't be updated, no point in passing them.

File: frontend/client/components/Profile/index.tsx [hunk 3]
```diff
@@ -76,7 +78,7 @@ class Profile extends React.Component<Props> {
           description={`Join ${user.name} in funding the future!`}
           image={user.avatarUrl}
         />
-        <ProfileUser user={user} />
+        <ProfileUser user={user} isSelf={isSelf} onEdit={this.props.updateUser} />
```
Comment (wbobeirne) [inline frontend/client/components/Profile/index.tsx]: Opinion: ProfileUser should be `connect`ed to determine isSelf & connect updateUser itself.

File: frontend/client/utils/helpers.ts [hunk 4]
```diff
@@ -26,6 +27,14 @@ export function findComment(
   return null;
 }
 
+// clone and filter keys by keySource object's keys
+export function cleanClone<T extends object>(keySource: T, target: object) {
+  const sourceKeys = Object.keys(keySource);
+  const fullClone = lodash.clone(target);
```
Comment (wbobeirne) [inline frontend/client/utils/helpers.ts]: Prefer `{ ...target }`, lodash's clone is a crazy big function that's a bit overkill. Unless I'm missing something that it does for this case.

File: frontend/client/utils/helpers.ts [hunk 5]
```diff
@@ -26,6 +27,14 @@ export function findComment(
   return null;
 }
 
+// clone and filter keys by keySource object's keys
+export function cleanClone<T extends object>(keySource: T, target: object) {
```
Comment (wbobeirne) [inline frontend/client/utils/helpers.ts]: `target` can be `Partial<T>` here. This is kinda cool cause TS will throw if target doesn't have _any_ keys that overlap with source, or if one of the keys is the wrong type.

File: frontend/client/utils/helpers.ts (hunk 4 above)
Comment (AMStrix) [reply frontend/client/utils/helpers.ts]: Good point.

**Pull request merged**